Sprint 3

MasterOfMenace opened the pull request.

Commit:
Приводит в порядок отступы в шаблонах

Commit:
Переносит скрипты в head

Commit:
Выносит цвета в css переменные

Commit:
Добавляет дженерик в Block

Commit:
Уточняет типы пропстов у копмпонентов

Commit:
Исправляет замечания ревью

Commit:
Исправляет замечания ревью

Commit:
Созданы классы Route и Router

Commit:
Правка роутера и роутов. Написан index.ts.
Рефакторинг компонентов страниц mainPage, signUpPage, loginPage
Подключены к роутеру страницы mainPage, signUpPage, loginPage.
Написан хок withRouter

Commit:
Подключает остальные страницы к роутеру

Commit:
Добавлены классы BaseApi, Store
Добавлены функции в utils - merge, set

Commit:
ДОбавлены апи для автоизации. Созданы контроллеры
для страницы входа и страницы регистрации.

Commit:
добавлен логаут и получение данных текущего
пользователя

Commit:
Удалено добавление обработчиков из шаблонизатора

Commit:
ДОбавлена функция connect

Commit:
ДОбавлена страница для просмотра сведений
о пользователе, правка страницы для изменения сведений о пользователе.
Добавлен компонент userProfile. Реализовано отображения данных, полученных
с сервера

Commit:
Реализовано изменение информации пользователя

Commit:
Ссылки с иконкой назад вынесены из шаблонов,
создаются компонентом ссылки и переходят на шаг назад в истории

Commit:
Создан компонент ChatList.
Реализована загрузка списка чатов. Переработана логика отображения списка чатов.
Стили для страницы чатов вынесены в отдельный файл

Commit:
Выбор активного чата, отображение данных
пользователя на странице чата

Commit:
Чат вынесен в отдельный компонент.
Создан компонент пустого чата.

Commit:
Добавлен вебсокет.
Реализация отправки и получения сообщений

Commit:
Переработаны компоненты для уменьшения перерисовок

Commit:
Добавлено отображение плейсхолдера если нет
аватара и отображение аватара чата в списке чатов

Commit:
Добавлена подстановка плйесхолдера в сам компонент аватара,
добавлено оторажение аватара чата в текущем чата

Commit:
Добавлено отображение аватаров в чате

Commit:
Добавлено изменение аватара пользователя

Commit:
Добавлено отображение аватара текущего
ползователя в шапке чата, отображение аватара в сообщениях,
изменение положения сообщения для текущего пользователя

Commit:
добавлено модальное окно для создания чата

Commit:
Изменил порядок вызоыва метода componentDidUpdate

Commit:
Добавлена возможность создания чата

Commit:
Поправил стили модалки создания чата

Commit:
Добавлена иконка в кнопку добавления чата.
Удален лишний компонент.
Правка пропов className

Commit:
добавлен тип стора, добавлено обновление списка
чатов при отправке сообщения

Commit:
Добавлено модальное окно для поиска и выбора
пльзователя

Commit:
Реализован поиск и добавление пользователя в чат

Commit:
Добавляет chai, mocha и тест для шаблонизатора

Commit:
Тесты для httpTransport и templator

Commit:
Тесты для роутера и блока. Правки настроек бабеля
для тестов. Установлен пакет cross-env

Commit:
Добавлено отображение времени сообщений

Commit:
Добавлен контроллер mainPage.
Добавлен редирект зарегистрированного и незарегистрированного пользователя

Commit:
ДОбавлен переход на страницу настроек пользователя

Commit:
Правит скрипт, удаляет ненужную статику

Commit:
Устанавливает babel/core, правит стили списка
сообщений

Commit:
Правит настройки stylelint.
Отключена проверка на дубликаты правил, так как после автопрефиксера
падают автотесты.

Commit:
Дополняет README

Commit:
ДОбавляет ссылку на netlify

dmitrymorozoff commented on src/AddUserModal/AddUserModal.ts:
```diff
@@ -0,0 +1,139 @@
+  onSelectUser?: (id: number) => void;
+};
+
+const userApi = new UserApi();
```
UserApi можно сделать синглтоном, и экспортировать инстанс прямо из файла с UserApi

MasterOfMenace replied:
сделал

dmitrymorozoff commented on src/AddUserModal/addUserModal.tmpl.ts:
```diff
@@ -0,0 +1,16 @@
+  <div class="modal__inner">
+    <div class="add-chat-user-modal">
+      <p class="add-chat-user-modal__title">Добавить пользователя</p>
+      <!--<form class="add-chat-user-modal__form">
```
Комментарии лучше удалять все

MasterOfMenace replied:
удалил

dmitrymorozoff commented on src/api/authApi/AuthApi.ts:
```diff
@@ -0,0 +1,69 @@
+  password: string;
+};
+
+const authHttpTransport = new HTTPTransport(BASE_URL);
```
HTTPTransport можно инициализировать внутри класса AuthApi

MasterOfMenace replied:
убрал в поле класса, инициализацию перенес в конструктор

dmitrymorozoff commented on src/api/authApi/AuthApi.ts:
```diff
@@ -0,0 +1,69 @@
+export class AuthApi extends BaseApi {
+  login(data: {login: string; password: string}) {
+    return authHttpTransport
+      .post('/auth/signin', {
```
`/auth/` лучше тоже добавить в базовый урл

MasterOfMenace replied:
добавил

dmitrymorozoff commented on src/api/authApi/AuthApi.ts:
```diff
@@ -0,0 +1,69 @@
+    return authHttpTransport
+      .post('/auth/signin', {
+        data,
+        headers: {
```
Чтобы не дублировать везде заголовки, можно вторым параметром в HTTPTransport добавить набор базовых опций, а уже в самих методах post, get ... дать возможность менять эти опции

dmitrymorozoff commented on src/api/baseApi/BaseApi.ts:
```diff
@@ -0,0 +1,25 @@
+export abstract class BaseApi {
```
От этого класса нет никакого смысла, он по сути ничего не делает, и наследоваться от него не нужно. Лучше удалить

MasterOfMenace replied:
удалил

dmitrymorozoff commented on src/components/avatar/Avatar.ts:
```diff
@@ -1,15 +1,46 @@
+export default class Avatar extends Block<AvatarProps> {
   constructor(props: AvatarProps) {
-    super('div', props);
+    const defaultProps: Partial<AvatarProps> = {
```
Смысла в доп объекте нет, можно сразу писать `props.avatarSrc || imagePlaceholder`

MasterOfMenace replied:
убрал объект

dmitrymorozoff commented on src/components/chatList/ChatList.ts:
```diff
@@ -0,0 +1,78 @@
+import template from './chatlist.tmpl';
+
+/*
+[
```
Что-то лишнее

MasterOfMenace replied:
удалил

dmitrymorozoff commented on src/components/block/Block.ts:
```diff
@@ -11,17 +11,32 @@ export type EventType = {
 };
 
-class Block {
+export interface IBlock<T extends Props = Props> {
```
Не понял зачем этот интерфейс, он же нигде не используется

MasterOfMenace replied:
удалил

dmitrymorozoff commented on src/components/message/message.tmpl.ts:
```diff
@@ -1,13 +1,12 @@
+<li class="messages-list__item message {{className}}">
   {{avatar}}
   <div class="message__message">
     <p class="message__text">
```
Для времени можно использовать тег time

dmitrymorozoff commented on src/pages/chatPage/chatPage.ts:
```diff
@@ -2,159 +2,175 @@ import Block from '../../components/block/Block';
+    const addChatButton = new Button({
+      type: 'button',
+      text: `
+      <svg width="24" height="24" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
```
Иконку лучше конечно было бы вынести в отдельный шаблон, очень много места она тут занимает

dmitrymorozoff commented on src/pages/loginPage/loginPage.controller.ts:
```diff
@@ -0,0 +1,21 @@
+  public static login(data: {login: string; password: string}) {
+    authApi
+      .login(data)
+      .then(() => authApi.getUserData())
```
Вместо промисов , можно посмотреть в сторону [async/await](https://learn.javascript.ru/async-await). Запись будет выглядеть красивее и читабельнее

MasterOfMenace replied:
буду иметь в виду, спасибо

dmitrymorozoff commented on src/pages/userSettingsShow/userSettingsShow.ts:
```diff
@@ -0,0 +1,157 @@
+  phone: '',
+});
+
+const userInfo = new UserInfo({
```
Инициализацию лучше перенести в сам компонент

dmitrymorozoff commented on src/store/connect.ts:
```diff
@@ -0,0 +1,55 @@
+import store, {STORE_EVENTS} from './Store';
+
+/**
+ * function connect(Component: typeof Block, mapStateToProps: (state: Indexed) => Indexed) {
```
Лучше удалять весь закомментированный код, в случае необходимости его легко можно найти в истории гит

MasterOfMenace replied:
исправлено

dmitrymorozoff commented on src/utils/MessagesTransport.ts:
```diff
@@ -0,0 +1,92 @@
+import store from '../store/Store';
+import {isObject} from './utils';
+
+function isMessagesArrayData(data: unknown): data is ChatMessage[] {
```
Вспомогательные утилиты лучше вынести в отдельные файлы, чтобы их легко было найти ну и не засорять этот класс

MasterOfMenace replied:
вынес наружу из модуля в отдельный файл

dmitrymorozoff commented on src/utils/MessagesTransport.ts:
```diff
@@ -0,0 +1,92 @@
+}
+
+export class MessagesTransport extends EventBus {
+  EVENTS = {
```
Лучше вынести в enum все события

MasterOfMenace replied:
@dmitrymorozoff
Вынести вообще все события в один энам или для каждого класса где они нужны? И почему лучше энам, а не константа?  Сами разработчики ТС пишут что энамы можно не использовать, а использовать объекты https://www.typescriptlang.org/docs/handbook/enums.html#objects-vs-enums
Или речь о том, чтобы не держать в поле класса?

dmitrymorozoff replied:
Да, события не должны находиться внутри класса MessagesTransport. Можно в const, можно в const enum.

dmitrymorozoff commented on src/utils/utils.ts:
```diff
@@ -54,3 +57,94 @@ export function getDeepCopy<T>(object: T): T {
+  return lhs;
+}
+
+export function set<T extends Indexed>(object: T, path: string, value: unknown): T {
```
Если функция в файле станет еще больше, будет затруднительно его листать и искать нужную утилиту. Лучше каждую утилиту вынести в свой отдельный файл

MasterOfMenace replied:
Разнес все утилиты по отдельным файлам

dmitrymorozoff reviewed (commented):
Привет) Спасибо за выполненную работу. В целом все неплохо, все задачи ТЗ выполнены, оставил несколько комментариев, которые можно было бы поправить. 

Из моментов, которые нужно поправить перед следующим спринтом:

- Поскольку роутер клиентский - нужно указать netlify, чтобы он на любой запрос отдавал index.html. Это настройки netlify.toml. https://docs.netlify.com/routing/redirects/
```
...
[[redirects]]
  from = "*"
  to = "/"
  status = 200
```
- после отправки сообщения, хорошо бы подчищать поле ввода
- когда создаешь чат, почему то в заголовке слева пишется `NaNNaN /span`

Commit:
ДОбавляет переадресацию на иднекс для netlify.
Удаляет класс BaseApi

Commit:
Устранение замечаний ревью.
Выносит функции из утилс по отдельным папкам.
Исправляет некорректную отрисовку элементов списка чатов.
Добавляет очищение поля ввода после отправки сообщения.
Разносит по отдельным директориям Templator, MessagesTransport, HTTPTransport.

Commit:
Удаляет редирект для нетлифай

Commit:
Возвращает уточненную нстройку редирект в netlify

dmitrymorozoff commented:
Привет, вижу что были правки, но на нетлифай сейчас старая версия, где нет никаких правок. Проверь, пожалуйста

dmitrymorozoff commented:
Все хорошо, спасибо за правки)

MasterOfMenace committed:
Merge branch 'main' into sprint_3
